### PR TITLE
[Woptim] Compiler wrappers + update packages + use spack environments

### PR DESCRIPTION
**Pull request by adrienbernede**
- [x] What should we do with the content of our local modules.yaml file?
So far the modules.yaml files hasn’t been merged in the spack.yaml files. It seems to work fine.

## Summary

- The compilers (wrappers) on Lassen have been updated to better support the use of gcc 8.3.1 toolchain.
- The local spack packages have been updated to be as close as possible to what we are pushing in https://github.com/spack/spack/pull/41375.
- We moved to Spack environments, allowing us to update Uberenv in the projects.

## Notes

Packages in RADIUSS Spack Configs still have some specific changes compared to the PR mentioned above.

## Tests

https://github.com/LLNL/Umpire/pull/875
https://github.com/LLNL/RAJA/pull/1617
https://github.com/LLNL/RAJAPerf/pull/438
https://github.com/LLNL/Caliper/pull/520
https://github.com/LLNL/CHAI/pull/255

### CODE REVIEW

**Comment by adrienbernede** [comment]
@rhornung67 this is ready for review.

File: gitlab/radiuss-jobs/lassen.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@
 clang_12_0_1_ibm_gcc_8_3_1_cuda_10_1_243:
   variables:
     SPEC: "${PROJECT_LASSEN_VARIANTS} +cuda %clang@=12.0.1.ibm.gcc.8.3.1 ^cuda@10.1.243+allow-unsupported-compilers ${PROJECT_LASSEN_DEPS}"
+    MODULE_LIST: "cuda/10.1.243"
```
Comment (rhornung67) [inline gitlab/radiuss-jobs/lassen.yml:11]: To make choosing jobs easier for projects, would it be better to only have jobs corresponding to system defaults for compilers and packages?  For example, CUDA 11.2.0 is the default now on LC blueos machines and there may be only one project, if any still using a CUDA 10.x version.

In other words, if you want to test anything other than system defaults I can add it to project file. If it's in here, I have to turn it off.

File: packages/raja/package.py [hunk 2]
```diff
@@ -1,60 +1,147 @@
...
-    version("0.6.0", tag="v0.6.0", submodules="True")
-    version("0.5.3", tag="v0.5.3", submodules="True")
-    version("0.5.2", tag="v0.5.2", submodules="True")
-    version("0.5.1", tag="v0.5.1", submodules="True")
-    version("0.5.0", tag="v0.5.0", submodules="True")
-    version("0.4.1", tag="v0.4.1", submodules="True")
-    version("0.4.0", tag="v0.4.0", submodules="True")
+    version("develop", branch="develop", submodules=submodules)
+    version("main", branch="main", submodules=submodules)
+    version(
```
Comment (rhornung67) [inline packages/raja/package.py:37]: v2024.02.1 release is out

File: toss_3_x86_64_ib/spack.yaml [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+      - $user_cache_path/stage
+    test_stage: $user_cache_path/test
+    misc_cache: $user_cache_path/misc
+    flags:
+      keep_werror: 'all'
+
+    view: false
+
+  compilers::
+  - compiler:
```
Comment (rhornung67) [inline toss_3_x86_64_ib/spack.yaml:30]: Similar question to the one I asked about old versions in p[roject Spack packages: should we remove old compiler versions that are no longer available?  For example, on TOSS4 systems, clang/14.0.6 appears to be the only version available now.

File: toss_4_x86_64_ib_cray/spack.yaml [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    build_stage::
+      - $user_cache_path/stage
+    test_stage: $user_cache_path/test
+    misc_cache: $user_cache_path/misc
+    flags:
+      keep_werror: 'all'
+
+    view: false
+
+  compilers::
```
Comment (rhornung67) [inline toss_4_x86_64_ib_cray/spack.yaml:29]: Similar question to earlier one: cce/14.0.3 is the oldest version available on these LC systems.

**Review by rhornung67 — APPROVED** [review]
Overall, this looks OK. We should double-check which compiler versions are actually available on LC systems and whether we want to remove versions that aren't.

Also, a similar question about project versions included in project Spack packages.

File: gitlab/radiuss-jobs/lassen.yml (hunk 1 above)
Comment (adrienbernede) [reply gitlab/radiuss-jobs/lassen.yml:11]: The rationale for choosing those jobs was to list was projects need in production. If the needs have changed, we just need to change those spec accordingly.
This addition is just to load the appropriate module for the spec.

Comment (rhornung67) [reply gitlab/radiuss-jobs/lassen.yml:11]: I'll start an issue here to collect compilers and versions projects use in production. Hopefully, we can then agree on what we want in the shared specs.

File: packages/raja/package.py [hunk 5]
```diff
@@ -1,60 +1,147 @@
...
+        commit="be91e040130678b1350dbda56cc352433db758bd",
+        submodules="True",
+    )
+    version(
+        "0.10.0",
+        tag="v0.10.0",
+        commit="53cb89cf788d28bc4ed2b4e6f75483fdd26024aa",
+        submodules="True",
+    )
+    version(
```
[the remark this answers is not in the document]
Comment (adrienbernede) [reply packages/raja/package.py:115]: My impression is that packages tend to keep old versions. I don’t know of a convention.

[the remark this answers is not in the document]
Comment (rhornung67) [reply packages/raja/package.py:115]: That's fine. I was just being curious.

[the remark this answers is not in the document]
Comment (adrienbernede) [reply packages/raja/package.py:115]: Maybe you should ask on Spack MS Teams channel.

File: toss_3_x86_64_ib/spack.yaml (hunk 3 above)
Comment (adrienbernede) [reply toss_3_x86_64_ib/spack.yaml:30]: We can definitely do some clean up here. Let me open an issue for that.

File: toss_4_x86_64_ib_cray/spack.yaml (hunk 4 above)
Comment (adrienbernede) [reply toss_4_x86_64_ib_cray/spack.yaml:29]: I’m gonna treat this demand in a subsequent PR. The present one already has enough changes in it.

**Pull request merged**